File: app/src/main/java/dev/ragnarok/fenrir/upload/UploadManagerImpl.java
package dev.ragnarok.fenrir.upload;

import static dev.ragnarok.fenrir.util.Objects.isNull;
import static dev.ragnarok.fenrir.util.Objects.nonNull;
import static dev.ragnarok.fenrir.util.Utils.firstNonEmptyString;
import static dev.ragnarok.fenrir.util.Utils.getCauseIfRuntime;
import static dev.ragnarok.fenrir.util.Utils.nonEmpty;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.os.Build;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import dev.ragnarok.fenrir.Extra;
import dev.ragnarok.fenrir.Injection;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.api.ApiException;
import dev.ragnarok.fenrir.api.PercentagePublisher;
import dev.ragnarok.fenrir.api.interfaces.INetworker;
import dev.ragnarok.fenrir.api.model.server.UploadServer;
import dev.ragnarok.fenrir.db.interfaces.IStorages;
import dev.ragnarok.fenrir.domain.IAttachmentsRepository;
import dev.ragnarok.fenrir.domain.IWallsRepository;
import dev.ragnarok.fenrir.longpoll.NotificationHelper;
import dev.ragnarok.fenrir.service.ErrorLocalizer;
import dev.ragnarok.fenrir.upload.impl.AudioUploadable;
import dev.ragnarok.fenrir.upload.impl.ChatPhotoUploadable;
import dev.ragnarok.fenrir.upload.impl.DocumentUploadable;
import dev.ragnarok.fenrir.upload.impl.OwnerPhotoUploadable;
import dev.ragnarok.fenrir.upload.impl.Photo2AlbumUploadable;
import dev.ragnarok.fenrir.upload.impl.Photo2MessageUploadable;
import dev.ragnarok.fenrir.upload.impl.Photo2WallUploadable;
import dev.ragnarok.fenrir.upload.impl.RemoteAudioPlayUploadable;
import dev.ragnarok.fenrir.upload.impl.StoryUploadable;
import dev.ragnarok.fenrir.upload.impl.Video2WallUploadable;
import dev.ragnarok.fenrir.upload.impl.VideoToMessageUploadable;
import dev.ragnarok.fenrir.upload.impl.VideoUploadable;
import dev.ragnarok.fenrir.util.Optional;
import dev.ragnarok.fenrir.util.Pair;
import dev.ragnarok.fenrir.util.Utils;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.processors.PublishProcessor;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class UploadManagerImpl implements IUploadManager {

    private static final int PROGRESS_LOOKUP_DELAY = 500;
    private static final String NOTIFICATION_CHANNEL_ID = "upload_files";

    private final Context context;
    private final INetworker networker;
    private final IStorages storages;
    private final IAttachmentsRepository attachmentsRepository;
    private final IWallsRepository walls;
    private final List<Upload> queue = new ArrayList<>();
    private final Scheduler scheduler;

    private final PublishProcessor<List<Upload>> addingProcessor = PublishProcessor.create();
    private final PublishProcessor<int[]> deletingProcessor = PublishProcessor.create();
    private final PublishProcessor<Pair<Upload, UploadResult<?>>> completeProcessor = PublishProcessor.create();
    private final PublishProcessor<Upload> statusProcessor = PublishProcessor.create();

    private final Flowable<Long> timer;
    private final CompositeDisposable notificationUpdateDisposable = new CompositeDisposable();
    private final Map<String, UploadServer> serverMap = Collections.synchronizedMap(new HashMap<>());
    private final CompositeDisposable compositeDisposable = new CompositeDisposable();
    private volatile Upload current;
    private boolean needCreateChannel = true;

    public UploadManagerImpl(Context context, INetworker networker, IStorages storages, IAttachmentsRepository attachmentsRepository,
                             IWallsRepository walls) {
        this.context = context.getApplicationContext();
        this.networker = networker;
        this.storages = storages;
        this.attachmentsRepository = attachmentsRepository;
        this.walls = walls;
        scheduler = Schedulers.from(Executors.newSingleThreadExecutor());
        timer = Flowable.interval(PROGRESS_LOOKUP_DELAY, PROGRESS_LOOKUP_DELAY, TimeUnit.MILLISECONDS).onBackpressureBuffer();
    }

    private static Upload intent2Upload(UploadIntent intent) {
        return new Upload(intent.getAccountId())
                .setAutoCommit(intent.isAutoCommit())
                .setDestination(intent.getDestination())
                .setFileId(intent.getFileId())
                .setFileUri(intent.getFileUri())
                .setStatus(Upload.STATUS_QUEUE)
                .setSize(intent.getSize());
    }

    private static String createServerKey(Upload upload) {
        UploadDestination dest = upload.getDestination();

        StringBuilder builder = new StringBuilder();
        builder.append(Extra.ACCOUNT_ID).append(upload.getAccountId());
        builder.append(Extra.METHOD).append(dest.getMethod());

        switch (upload.getDestination().getMethod()) {
            case Method.DOCUMENT:
            case Method.VIDEO:
            case Method.AUDIO:
            case Method.TO_COMMENT:
            case Method.TO_WALL:
                if (dest.getOwnerId() < 0) {
                    builder.append(Extra.GROUP_ID).append(Math.abs(dest.getOwnerId()));
                }
                break;
            case Method.PHOTO_TO_ALBUM:
                builder.append(Extra.ALBUM_ID).append(dest.getId());
                if (dest.getOwnerId() < 0) {
                    builder.append(Extra.GROUP_ID).append(Math.abs(dest.getOwnerId()));
                }
                break;
            case Method.STORY:
            case Method.REMOTE_PLAY_AUDIO:
            case Method.TO_MESSAGE:
                //do nothink
                break;
            case Method.PHOTO_TO_PROFILE:
            case Method.PHOTO_TO_CHAT:
                builder.append(Extra.OWNER_ID).append(dest.getOwnerId());
                break;
        }

        return builder.toString();
    }

    @Override
    public Single<List<Upload>> get(int accountId, @NonNull UploadDestination destination) {
        return Single.fromCallable(() -> getByDestination(accountId, destination));
    }

    private List<Upload> getByDestination(int accountId, @NonNull UploadDestination destination) {
        synchronized (this) {
            List<Upload> data = new ArrayList<>();
            for (Upload upload : queue) {
                if (accountId == upload.getAccountId() && destination.compareTo(upload.getDestination())) {
                    data.add(upload);
                }
            }
            return data;
        }
    }

    private void startWithNotification() {
        updateNotification(Collections.emptyList());

        notificationUpdateDisposable.add(observeProgress()
                .observeOn(Injection.provideMainThreadScheduler())
                .subscribe(this::updateNotification));
    }

    private void updateNotification(List<IProgressUpdate> updates) {
        if (nonEmpty(updates)) {
            int progress = updates.get(0).getProgress();

            NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
            if (isNull(notificationManager)) {
                return;
            }

            NotificationCompat.Builder builder;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                if (needCreateChannel) {
                    NotificationChannel channel = new NotificationChannel(NOTIFICATION_CHANNEL_ID, context.getString(R.string.channel_upload_files), NotificationManager.IMPORTANCE_LOW);
                    notificationManager.createNotificationChannel(channel);
                    needCreateChannel = false;
                }

                builder = new NotificationCompat.Builder(context, NOTIFICATION_CHANNEL_ID);
            } else {
                builder = new NotificationCompat.Builder(context, NOTIFICATION_CHANNEL_ID).setPriority(Notification.PRIORITY_LOW);
            }

            builder.setContentTitle(context.getString(R.string.files_uploading_notification_title))
                    .setSmallIcon(R.drawable.ic_notification_upload)
                    .setOngoing(true)
                    .setProgress(100, progress, false)
                    .build();

            notificationManager.notify(NotificationHelper.NOTIFICATION_UPLOAD, builder.build());
        }
    }

    private void stopNotification() {
        notificationUpdateDisposable.clear();
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (nonNull(notificationManager)) {
            notificationManager.cancel(NotificationHelper.NOTIFICATION_UPLOAD);
        }
    }

    @Override
    public void enqueue(@NonNull List<UploadIntent> intents) {
        synchronized (this) {
            List<Upload> all = new ArrayList<>(intents.size());

            for (UploadIntent intent : intents) {
                Upload upload = intent2Upload(intent);
                all.add(upload);
                queue.add(upload);
            }

            addingProcessor.onNext(all);
            startIfNotStarted();
        }
    }

    private void startIfNotStarted() {
        compositeDisposable.add(Completable.complete()
                .observeOn(scheduler)
                .subscribe(this::startIfNotStartedInternal));
    }

    private Upload findFirstQueue() {
        Upload first = null;
        for (Upload u : queue) {
            if (u.getStatus() == Upload.STATUS_QUEUE) {
                first = u;
                break;
            }
        }
        return first;
    }

    private void startIfNotStartedInternal() {
        synchronized (this) {
            if (current != null) {
                return;
            }

            Upload first = findFirstQueue();
            if (first == null) {
                stopNotification();
                return;
            }

            startWithNotification();

            current = first;

            first.setStatus(Upload.STATUS_UPLOADING).setErrorText(null);
            statusProcessor.onNext(first);

            IUploadable<?> uploadable = createUploadable(first);
            UploadServer server = serverMap.get(createServerKey(first));

            compositeDisposable.add(uploadable.doUpload(first, server, new WeakProgressPublisgher(first))
                    .subscribeOn(scheduler)
                    .observeOn(scheduler)
                    .subscribe(result -> onUploadComplete(first, result), t -> onUploadFail(first, t)));
        }
    }

    private void onUploadComplete(Upload upload, UploadResult<?> result) {
        synchronized (this) {
            queue.remove(upload);

            if (current == upload) {
                current = null;
            }

            //final int accountId = upload.getAccountId();
            //final UploadDestination destination = upload.getDestination();
            //if (destination.getMethod() == Method.TO_MESSAGE && getByDestination(accountId, destination).isEmpty()) {
            //    sendMessageIfWaitForUpload(accountId, destination.getId());
            //}

            UploadDestination destination = upload.getDestination();
            if (destination.getMessageMethod() != MessageMethod.VIDEO && destination.getMethod() != Method.VIDEO && destination.getMethod() != Method.STORY)
                serverMap.put(createServerKey(upload), result.getServer());

            completeProcessor.onNext(Pair.Companion.create(upload, result));
            startIfNotStartedInternal();
        }
    }

    private void onUploadFail(Upload upload, Throwable t) {
        synchronized (this) {
            if (current == upload) {
                current = null;

                Throwable cause = getCauseIfRuntime(t);
                String message;
                if (cause instanceof ApiException) {
                    message = ErrorLocalizer.localizeThrowable(context, cause);
                } else {
                    message = firstNonEmptyString(cause.getMessage(), cause.toString());
                }
                compositeDisposable.add(Completable.complete()
                        .observeOn(Injection.provideMainThreadScheduler())
                        .subscribe(() -> Toast.makeText(context, message, Toast.LENGTH_SHORT).show()));

            }
            String errorMessage;
            if (t instanceof ApiException) {
                errorMessage = ErrorLocalizer.localizeThrowable(context, t);
            } else {
                errorMessage = firstNonEmptyString(t.getMessage(), t.toString());
            }
            upload.setStatus(Upload.STATUS_ERROR).setErrorText(errorMessage);
            statusProcessor.onNext(upload);

            startIfNotStartedInternal();
        }
    }

    @Override
    public void cancel(int id) {
        synchronized (this) {
            if (current != null && current.getId() == id) {
                compositeDisposable.clear();
                current = null;
            }

            int index = Utils.findIndexById(queue, id);
            if (index != -1) {
                queue.remove(index);
                deletingProcessor.onNext(new int[]{id});
            }

            startIfNotStarted();
        }
    }

    @Override
    public void cancelAll(int accountId, @NonNull UploadDestination destination) {
        synchronized (this) {
            if (current != null && accountId == current.getAccountId() && destination.compareTo(current.getDestination())) {
                compositeDisposable.clear();
                current = null;
            }

            List<Upload> target = new ArrayList<>();

            Iterator<Upload> iterator = queue.iterator();
            while (iterator.hasNext()) {
                Upload next = iterator.next();
                if (accountId == next.getAccountId() && destination.compareTo(next.getDestination())) {
                    iterator.remove();
                    target.add(next);
                }
            }

            if (!target.isEmpty()) {
                int[] ids = new int[target.size()];
                for (int i = 0; i < target.size(); i++) {
                    ids[i] = target.get(i).getId();
                }
                deletingProcessor.onNext(ids);
            }

            startIfNotStarted();
        }
    }

    @Override
    public Optional<Upload> getCurrent() {
        synchronized (this) {
            return Optional.wrap(current);
        }
    }

    @Override
    public Flowable<int[]> observeDeleting(boolean includeCompleted) {
        if (includeCompleted) {
            Flowable<int[]> completeIds = completeProcessor.onBackpressureBuffer()
                    .map(pair -> new int[]{pair.getFirst().getId()});

            return Flowable.merge(deletingProcessor.onBackpressureBuffer(), completeIds);
        }

        return deletingProcessor.onBackpressureBuffer();
    }

    @Override
    public Flowable<List<Upload>> observeAdding() {
        return addingProcessor.onBackpressureBuffer();
    }

    @Override
    public Flowable<Upload> obseveStatus() {
        return statusProcessor.onBackpressureBuffer();
    }

    @Override
    public Flowable<Pair<Upload, UploadResult<?>>> observeResults() {
        return completeProcessor.onBackpressureBuffer();
    }

    @Override
    public Flowable<List<IProgressUpdate>> observeProgress() {
        return timer.map(ignored -> {
            synchronized (this) {
                if (current == null) {
                    return Collections.emptyList();
                }

                IProgressUpdate update = new ProgressUpdate(current.getId(), current.getProgress());
                return Collections.singletonList(update);
            }
        });
    }

    private IUploadable<?> createUploadable(Upload upload) {
        UploadDestination destination = upload.getDestination();

        switch (destination.getMethod()) {
            case Method.VIDEO:
                return new VideoUploadable(context, networker);
            case Method.STORY:
                return new StoryUploadable(context, networker);
            case Method.AUDIO:
                return new AudioUploadable(context, networker);
            case Method.REMOTE_PLAY_AUDIO:
                return new RemoteAudioPlayUploadable(context, networker);
            case Method.TO_MESSAGE:
                if (destination.getMessageMethod() == MessageMethod.PHOTO)
                    return new Photo2MessageUploadable(context, networker, attachmentsRepository, storages.messages());
                else if (destination.getMessageMethod() == MessageMethod.VIDEO)
                    return new VideoToMessageUploadable(context, networker, attachmentsRepository, storages.messages());
                else
                    throw new UnsupportedOperationException();
            case Method.PHOTO_TO_ALBUM:
                return new Photo2AlbumUploadable(context, networker, storages.photos());
            case Method.DOCUMENT:
                return new DocumentUploadable(context, networker, storages.docs());
            case Method.TO_COMMENT:
            case Method.TO_WALL:
                if (destination.getMessageMethod() == MessageMethod.PHOTO)
                    return new Photo2WallUploadable(context, networker, attachmentsRepository);
                else if (destination.getMessageMethod() == MessageMethod.VIDEO)
                    return new Video2WallUploadable(context, networker, attachmentsRepository);
                else
                    throw new UnsupportedOperationException();
            case Method.PHOTO_TO_PROFILE:
                return new OwnerPhotoUploadable(context, networker, walls);
            case Method.PHOTO_TO_CHAT:
                return new ChatPhotoUploadable(context, networker);
        }

        throw new UnsupportedOperationException();
    }

    private static final class WeakProgressPublisgher implements PercentagePublisher {

        final WeakReference<Upload> reference;

        WeakProgressPublisgher(Upload upload) {
            reference = new WeakReference<>(upload);
        }

        @Override
        public void onProgressChanged(int percentage) {
            Upload upload = reference.get();
            if (upload != null) {
                upload.setProgress(percentage);
            }
        }
    }

    private static final class ProgressUpdate implements IProgressUpdate {

        final int id;
        final int progress;

        private ProgressUpdate(int id, int progress) {
            this.id = id;
            this.progress = progress;
        }

        @Override
        public int getId() {
            return id;
        }

        @Override
        public int getProgress() {
            return progress;
        }
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/upload/UploadManagerImpl.kt
package dev.ragnarok.fenrir.upload

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.os.Build
import android.widget.Toast
import androidx.core.app.NotificationCompat
import dev.ragnarok.fenrir.Extra
import dev.ragnarok.fenrir.Includes.provideMainThreadScheduler
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.api.ApiException
import dev.ragnarok.fenrir.api.PercentagePublisher
import dev.ragnarok.fenrir.api.interfaces.INetworker
import dev.ragnarok.fenrir.api.model.server.UploadServer
import dev.ragnarok.fenrir.db.interfaces.IStorages
import dev.ragnarok.fenrir.domain.IAttachmentsRepository
import dev.ragnarok.fenrir.domain.IWallsRepository
import dev.ragnarok.fenrir.longpoll.NotificationHelper
import dev.ragnarok.fenrir.nonNullNoEmpty
import dev.ragnarok.fenrir.service.ErrorLocalizer.localizeThrowable
import dev.ragnarok.fenrir.upload.IUploadManager.IProgressUpdate
import dev.ragnarok.fenrir.upload.impl.*
import dev.ragnarok.fenrir.util.Optional
import dev.ragnarok.fenrir.util.Optional.Companion.wrap
import dev.ragnarok.fenrir.util.Pair
import dev.ragnarok.fenrir.util.Pair.Companion.create
import dev.ragnarok.fenrir.util.Utils.findIndexById
import dev.ragnarok.fenrir.util.Utils.firstNonEmptyString
import dev.ragnarok.fenrir.util.Utils.getCauseIfRuntime
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.processors.PublishProcessor
import io.reactivex.rxjava3.schedulers.Schedulers
import java.lang.ref.WeakReference
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import kotlin.math.abs

class UploadManagerImpl(
    context: Context,
    private val networker: INetworker,
    private val storages: IStorages,
    private val attachmentsRepository: IAttachmentsRepository,
    private val walls: IWallsRepository
) : IUploadManager {
    private val context: Context = context.applicationContext
    private val queue: MutableList<Upload> = ArrayList()
    private val scheduler: Scheduler = Schedulers.from(Executors.newSingleThreadExecutor())
    private val addingProcessor = PublishProcessor.create<List<Upload>>()
    private val deletingProcessor = PublishProcessor.create<IntArray>()
    private val completeProcessor = PublishProcessor.create<Pair<Upload, UploadResult<*>>>()
    private val statusProcessor = PublishProcessor.create<Upload>()
    private val timer: Flowable<Long> = Flowable.interval(
        PROGRESS_LOOKUP_DELAY.toLong(),
        PROGRESS_LOOKUP_DELAY.toLong(),
        TimeUnit.MILLISECONDS
    ).onBackpressureBuffer()
    private val notificationUpdateDisposable = CompositeDisposable()
    private val serverMap = Collections.synchronizedMap(HashMap<String, UploadServer>())
    private val compositeDisposable = CompositeDisposable()

    @Volatile
    private var current: Upload? = null
    private var needCreateChannel = true
    override fun get(accountId: Int, destination: UploadDestination): Single<List<Upload>> {
        return Single.fromCallable { getByDestination(accountId, destination) }
    }

    private fun getByDestination(accountId: Int, destination: UploadDestination): List<Upload> {
        synchronized(this) {
            val data: MutableList<Upload> = ArrayList()
            for (upload in queue) {
                if (accountId == upload.accountId && destination.compareTo(upload.destination)) {
                    data.add(upload)
                }
            }
            return data
        }
    }

    private fun startWithNotification() {
        updateNotification(emptyList())
        notificationUpdateDisposable.add(observeProgress()
            .observeOn(provideMainThreadScheduler())
            .subscribe { updateNotification(it) })
    }

    @Suppress("DEPRECATION")
    private fun updateNotification(updates: List<IProgressUpdate>) {
        if (updates.nonNullNoEmpty()) {
            val progress = updates[0].progress
            val notificationManager =
                context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager?
                    ?: return
            val builder: NotificationCompat.Builder
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                if (needCreateChannel) {
                    val channel = NotificationChannel(
                        NOTIFICATION_CHANNEL_ID,
                        context.getString(R.string.channel_upload_files),
                        NotificationManager.IMPORTANCE_LOW
                    )
                    notificationManager.createNotificationChannel(channel)
                    needCreateChannel = false
                }
                builder = NotificationCompat.Builder(context, NOTIFICATION_CHANNEL_ID)
            } else {
                builder = NotificationCompat.Builder(context, NOTIFICATION_CHANNEL_ID).setPriority(
                    Notification.PRIORITY_LOW
                )
            }
            builder.setContentTitle(context.getString(R.string.files_uploading_notification_title))
                .setSmallIcon(R.drawable.ic_notification_upload)
                .setOngoing(true)
                .setProgress(100, progress, false)
                .build()
            notificationManager.notify(NotificationHelper.NOTIFICATION_UPLOAD, builder.build())
        }
    }

    private fun stopNotification() {
        notificationUpdateDisposable.clear()
        val notificationManager =
            context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager?
        notificationManager?.cancel(NotificationHelper.NOTIFICATION_UPLOAD)
    }

    override fun enqueue(intents: List<UploadIntent>) {
        synchronized(this) {
            val all: MutableList<Upload> = ArrayList(intents.size)
            for (intent in intents) {
                val upload = intent2Upload(intent)
                all.add(upload)
                queue.add(upload)
            }
            addingProcessor.onNext(all)
            startIfNotStarted()
        }
    }

    private fun startIfNotStarted() {
        compositeDisposable.add(Completable.complete()
            .observeOn(scheduler)
            .subscribe { startIfNotStartedInternal() })
    }

    private fun findFirstQueue(): Upload? {
        var first: Upload? = null
        for (u in queue) {
            if (u.status == Upload.STATUS_QUEUE) {
                first = u
                break
            }
        }
        return first
    }

    private fun startIfNotStartedInternal() {
        synchronized(this) {
            if (current != null) {
                return
            }
            val first = findFirstQueue()
            if (first == null) {
                stopNotification()
                return
            }
            startWithNotification()
            current = first
            first.setStatus(Upload.STATUS_UPLOADING).errorText = null
            statusProcessor.onNext(first)
            val uploadable = createUploadable(first)
            val server = serverMap[createServerKey(first)]
            compositeDisposable.add(
                uploadable.doUpload(
                    first,
                    server,
                    WeakProgressPublisgher(first)
                )
                    .subscribeOn(scheduler)
                    .observeOn(scheduler)
                    .subscribe({
                        onUploadComplete(
                            first,
                            it
                        )
                    }, { t -> onUploadFail(first, t) })
            )
        }
    }

    private fun onUploadComplete(upload: Upload, result: UploadResult<*>) {
        synchronized(this) {
            queue.remove(upload)
            if (current === upload) {
                current = null
            }

            //final int accountId = upload.getAccountId();
            //final UploadDestination destination = upload.getDestination();
            //if (destination.getMethod() == Method.TO_MESSAGE && getByDestination(accountId, destination).isEmpty()) {
            //    sendMessageIfWaitForUpload(accountId, destination.getId());
            //}
            val destination = upload.destination
            if (destination.messageMethod != MessageMethod.VIDEO && destination.method != Method.VIDEO && destination.method != Method.STORY) serverMap[createServerKey(
                upload
            )] = result.server
            completeProcessor.onNext(create(upload, result))
            startIfNotStartedInternal()
        }
    }

    private fun onUploadFail(upload: Upload, t: Throwable) {
        synchronized(this) {
            if (current === upload) {
                current = null
                val cause = getCauseIfRuntime(t)
                val message: String? = if (cause is ApiException) {
                    localizeThrowable(context, cause)
                } else {
                    firstNonEmptyString(cause.message, cause.toString())
                }
                compositeDisposable.add(Completable.complete()
                    .observeOn(provideMainThreadScheduler())
                    .subscribe { Toast.makeText(context, message, Toast.LENGTH_SHORT).show() })
            }
            val errorMessage: String? = if (t is ApiException) {
                localizeThrowable(context, t)
            } else {
                firstNonEmptyString(t.message, t.toString())
            }
            upload.setStatus(Upload.STATUS_ERROR).errorText = errorMessage
            statusProcessor.onNext(upload)
            startIfNotStartedInternal()
        }
    }

    override fun cancel(id: Int) {
        synchronized(this) {
            if (current != null && (current ?: return@synchronized).id == id) {
                compositeDisposable.clear()
                current = null
            }
            val index = findIndexById(queue, id)
            if (index != -1) {
                queue.removeAt(index)
                deletingProcessor.onNext(intArrayOf(id))
            }
            startIfNotStarted()
        }
    }

    override fun cancelAll(accountId: Int, destination: UploadDestination) {
        synchronized(this) {
            if (current != null && accountId == current?.accountId && destination.compareTo(
                    current?.destination
                )
            ) {
                compositeDisposable.clear()
                current = null
            }
            val target: MutableList<Upload> = ArrayList()
            val iterator = queue.iterator()
            while (iterator.hasNext()) {
                val next = iterator.next()
                if (accountId == next.accountId && destination.compareTo(next.destination)) {
                    iterator.remove()
                    target.add(next)
                }
            }
            if (target.isNotEmpty()) {
                val ids = IntArray(target.size)
                for (i in target.indices) {
                    ids[i] = target[i].id
                }
                deletingProcessor.onNext(ids)
            }
            startIfNotStarted()
        }
    }

    override fun getCurrent(): Optional<Upload> {
        synchronized(this) { return wrap(current) }
    }

    override fun observeDeleting(includeCompleted: Boolean): Flowable<IntArray> {
        if (includeCompleted) {
            val completeIds = completeProcessor.onBackpressureBuffer()
                .map { intArrayOf(it.first.id) }
            return Flowable.merge(deletingProcessor.onBackpressureBuffer(), completeIds)
        }
        return deletingProcessor.onBackpressureBuffer()
    }

    override fun observeAdding(): Flowable<List<Upload>> {
        return addingProcessor.onBackpressureBuffer()
    }

    override fun obseveStatus(): Flowable<Upload> {
        return statusProcessor.onBackpressureBuffer()
    }

    override fun observeResults(): Flowable<Pair<Upload, UploadResult<*>>> {
        return completeProcessor.onBackpressureBuffer()
    }

    override fun observeProgress(): Flowable<List<IProgressUpdate>> {
        return timer.map {
            synchronized(this) {
                val pCurrent = current ?: return@map emptyList()
                val update: IProgressUpdate = ProgressUpdate(pCurrent.id, pCurrent.progress)
                return@map listOf(update)
            }
        }
    }

    private fun createUploadable(upload: Upload): IUploadable<*> {
        val destination = upload.destination
        when (destination.method) {
            Method.VIDEO -> return VideoUploadable(context, networker)
            Method.STORY -> return StoryUploadable(context, networker)
            Method.AUDIO -> return AudioUploadable(context, networker)
            Method.REMOTE_PLAY_AUDIO -> return RemoteAudioPlayUploadable(
                context,
                networker
            )
            Method.TO_MESSAGE -> return when (destination.messageMethod) {
                MessageMethod.PHOTO -> Photo2MessageUploadable(
                    context,
                    networker,
                    attachmentsRepository,
                    storages.messages()
                )
                MessageMethod.VIDEO -> VideoToMessageUploadable(
                    context,
                    networker,
                    attachmentsRepository,
                    storages.messages()
                )
                else -> throw UnsupportedOperationException()
            }
            Method.PHOTO_TO_ALBUM -> return Photo2AlbumUploadable(
                context,
                networker,
                storages.photos()
            )
            Method.DOCUMENT -> return DocumentUploadable(
                context,
                networker,
                storages.docs()
            )
            Method.TO_COMMENT, Method.TO_WALL -> return when (destination.messageMethod) {
                MessageMethod.PHOTO -> Photo2WallUploadable(
                    context,
                    networker,
                    attachmentsRepository
                )
                MessageMethod.VIDEO -> Video2WallUploadable(
                    context,
                    networker,
                    attachmentsRepository
                )
                else -> throw UnsupportedOperationException()
            }
            Method.PHOTO_TO_PROFILE -> return OwnerPhotoUploadable(
                context,
                networker,
                walls
            )
            Method.PHOTO_TO_CHAT -> return ChatPhotoUploadable(context, networker)
        }
        throw UnsupportedOperationException()
    }

    class WeakProgressPublisgher(upload: Upload) :
        PercentagePublisher {
        val reference: WeakReference<Upload> = WeakReference(upload)
        override fun onProgressChanged(percentage: Int) {
            val upload = reference.get()
            if (upload != null) {
                upload.progress = percentage
            }
        }

    }

    class ProgressUpdate(override val id: Int, override val progress: Int) : IProgressUpdate
    companion object {
        const val PROGRESS_LOOKUP_DELAY = 500
        const val NOTIFICATION_CHANNEL_ID = "upload_files"
        fun intent2Upload(intent: UploadIntent): Upload {
            return Upload(intent.accountId)
                .setAutoCommit(intent.pAutoCommit)
                .setDestination(intent.destination)
                .setFileId(intent.fileId)
                .setFileUri(intent.pFileUri)
                .setStatus(Upload.STATUS_QUEUE)
                .setSize(intent.size)
        }

        fun createServerKey(upload: Upload): String {
            val dest = upload.destination
            val builder = StringBuilder()
            builder.append(Extra.ACCOUNT_ID).append(upload.accountId)
            builder.append(Extra.METHOD).append(dest.method)
            when (upload.destination.method) {
                Method.DOCUMENT, Method.VIDEO, Method.AUDIO, Method.TO_COMMENT, Method.TO_WALL -> if (dest.ownerId < 0) {
                    builder.append(Extra.GROUP_ID).append(abs(dest.ownerId))
                }
                Method.PHOTO_TO_ALBUM -> {
                    builder.append(Extra.ALBUM_ID).append(dest.id)
                    if (dest.ownerId < 0) {
                        builder.append(Extra.GROUP_ID).append(abs(dest.ownerId))
                    }
                }
                Method.STORY, Method.REMOTE_PLAY_AUDIO, Method.TO_MESSAGE -> {}
                Method.PHOTO_TO_PROFILE, Method.PHOTO_TO_CHAT -> builder.append(
                    Extra.OWNER_ID
                ).append(dest.ownerId)
            }
            return builder.toString()
        }
    }

}